### PR TITLE
Convert an ink_release_assert into logic to reset the rbio to use the…

### PR DESCRIPTION
**Pull request by shinrich**
… remaining data.

### CODE REVIEW

File: iocore/net/SSLNetVConnection.cc [hunk 1]
```diff
@@ -512,13 +512,24 @@ SSLNetVConnection::net_read_io(NetHandler *nh, EThread *lthread)
...
-        free_handshake_buffers();
+      if (this->handShakeBuffer) {
+        if (BIO_eof(SSL_get_rbio(this->ssl)) && !handShakeReader->is_read_avail_more_than(0)) {
+          // Done with the buffer after the first exchange, convert over to the socket buffer
+          BIO *rbio = BIO_new_fd(this->get_socket(), BIO_NOCLOSE);
+          BIO_set_mem_eof_return(rbio, -1);
+          SSL_set0_rbio(this->ssl, rbio);
+          free_handshake_buffers();
+        } else if (handShakeReader->is_read_avail_more_than(0)) {
+          this->handShakeReader->consume(this->handShakeBioStored);
```
Comment (masaori335) [inline iocore/net/SSLNetVConnection.cc]: Is this block same logic in other 2 places (line 447 and line 1078) in this class?
If so, it'd be great if we can avoid copy & past.

Comment (shinrich) [inline iocore/net/SSLNetVConnection.cc]: Agreed.

**Comment by shinrich** [comment]
Thought on this some more this morning.  Perhaps all we should do is replace the ink_release_assert with a ink_assert and fail the connection in this case.  I think if the assert is triggered, the client is behaving badly.  Of course we can try to stumble on (this code addition).  Perhaps that is ok too.  But at a minimum the ink_release_assert should be removed/replaced.

**Comment by shinrich** [comment]
Pushed an updated PR that gathers up some of the cut-n-paste rbio code.  I'm testing this on one of our systems.

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by zwoop** [comment]
Testing on Docs / CI right now, once we land this on master (once reviewed), I can test on prod.

**Comment by shinrich** [comment]
It has been successfully running on my box today.  Of course that box was not seeing the release_assert before.

File: iocore/net/SSLNetVConnection.cc [hunk 2]
```diff
@@ -512,18 +530,15 @@ SSLNetVConnection::net_read_io(NetHandler *nh, EThread *lthread)
...
-        // Done with the buffer after the first exchange, convert over to the socket buffer
-        BIO *rbio = BIO_new_fd(this->get_socket(), BIO_NOCLOSE);
-        BIO_set_mem_eof_return(rbio, -1);
-        SSL_set0_rbio(this->ssl, rbio);
-        free_handshake_buffers();
+      if (this->handShakeBuffer) {
+        read.triggered = update_rbio(true);
       } else {
         Debug("ssl", "Want read from socket");
+        read.triggered = 0;
```
Comment (SolidWallOfCode) [inline iocore/net/SSLNetVConnection.cc]: This forces a blind tunnel to wait for the socket to be come read ready?

**Comment by SolidWallOfCode** [comment]
Nice cleanup of all that cut and paste stuff.

File: iocore/net/SSLNetVConnection.cc (hunk 2 above)
Comment (shinrich) [inline iocore/net/SSLNetVConnection.cc]: That is correct.  When the epoll shows data ready to read it will set the trigger.

**Pull request merged**